src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborWriter.cs

```diff
@@ -17,22 +19,33 @@ internal partial class CborWriter : IDisposable
         private byte[] _buffer = null!;
         private int _offset = 0;
 
+        private Stack<StackFrame>? _nestedDataItems;
+
         // remaining number of data items in current cbor context
         // with null representing indefinite length data items.
         // The root context ony permits one data item to be written.
-        private uint? _remainingDataItems = 1;
-        private bool _isEvenNumberOfDataItemsWritten = true; // required for indefinite-length map writes
-        private Stack<(CborMajorType type, bool isEvenNumberOfDataItemsWritten, uint? remainingDataItems)>? _nestedDataItemStack;
+        private int? _definiteLength = 1;
+        private int _itemsWritten = 0; // number of items written in the current context
+        private int _frameOffset = 0; // buffer offset particular to the current data item context
         private bool _isTagContext = false; // true if writer is expecting a tagged value
+        // Map-specific bookkeeping
+        private int? _currentKeyOffset = null;
+        private int? _currentValueOffset = null;
+        private SortedSet<(int Offset, int KeyLength, int TotalLength)>? _keyValueEncodingRanges = null;
 
-        public CborWriter()
+        public CborWriter(CborConformanceLevel conformanceLevel = CborConformanceLevel.Lax, bool patchIndefiniteLengthItems = false)
```

Comment:
This could perhaps use a configuration class, but it felt a bit premature with just two settings. Also note that patching defaults to false since it generally requires more work to support.

Reply:
I recommend changing `bool patchIndefiniteLengthItems = false` to `bool encodeIndefiniteLengths = false`, and throw in the ctor if the conformance level prohibits indefinite length encoding but it was requested.

* Validation of the incompatible state moves to the place the two parameters were specified, instead of arbitrarily later.
* The class generally just works.
* While the writer does more work, the reader stays on a likely better tested path (for arbitrary readers, not our implementation).
  * I don't know about home-spun CBOR implementations, but it's quite common for ASN.1-encoded data decoders to skip indefinite length support due to rarity.
* It uses the more conservative approach from section 3.1 ("Note that some applications and protocols will not want to use indefinite-length encoding.")

Reply:
> throw in the ctor if the conformance level prohibits indefinite length encoding but it was requested.

I like this, it will render a couple of failure modes impossible.

Reply:
Addressed.